daemon/cluster/executor/container/controller.go

```diff
@@ -217,17 +217,35 @@ func (r *controller) Start(ctx context.Context) error {
 
 				continue
 			}
-
+			// We should only remove networks after container networks are initialized.
+			// However, there are too many possible errors during container starting.
+			// Since removeNeworks() is an idempotent operation, it should be OK to removeNetworks
+			// here although it's not ideal.
+			log.G(ctx).Debugf("Removing networks after container %s failed to start.", r.adapter.container.name())
+			r.adapter.removeNetworks(ctx)
 			return errors.Wrap(err, "starting container failed")
 		}
 
 		break
 	}
 
+	// At this time, the container is started via containerd, and container.State.Running is true
+	// However, there is still possible errors from now on in this Start() function, and swarmkit will not
+	// call Shutdown() after task state is FAIL. We must clean up here.
```

Comment:
> and swarmkit will not call Shutdown() after task state is FAIL

Was this by design? Should this be changed? @dperny

Reply:
This is how current swarmkit works:
https://github.com/docker/swarmkit/blob/master/agent/exec/controller.go#L295
```
	if task.DesiredState >= api.TaskStateShutdown {
		if status.State >= api.TaskStateCompleted {
			return noop()
		}
...
```